Add analyzer option to limit the number of TUs imported

gamesh411 opened the pull request.

martong commented on lib/StaticAnalyzer/Core/AnalyzerOptions.cpp:
```diff
@@ -54,9 +54,9 @@ AnalyzerOptions::UserModeKind AnalyzerOptions::getUserMode() {
-      .Case("shallow", UMK_Shallow)
-      .Case("deep", UMK_Deep)
-      .Default(UMK_NotSet);
+                   .Case("shallow", UMK_Shallow)
```
Is this change related?

martong reviewed (commented):
You have so many unrelated changes which seems like the result of an automated formatting. Could you please avoid them?
You may use `git clang-format` which formats only the changed lines.

martong reviewed (approved):
Looks good!

gamesh411 commented:
I have run into a dependency issue. If I store the TU threshold as an AnalyzerOption, then I have to link with libStaticAnalyzerCore (see CMakeLists.txt change). But this introduces a cyclic dependency. I see a couple of solutions to this problem:

a) put the option elsewhere, for example as a general frontend-option
b) make the check elsewhere, for example inside some analyzer-specific manager class
c) try to cut the dependency by introducing intermediary libs

What are your thoughts?

martong commented:
> I have run into a dependency issue. If I store the TU threshold as an AnalyzerOption, then I have to link with libStaticAnalyzerCore (see CMakeLists.txt change). But this introduces a cyclic dependency. I see a couple of solutions to this problem:
> 
> a) put the option elsewhere, for example as a general frontend-option
> b) make the check elsewhere, for example inside some analyzer-specific manager class
> c) try to cut the dependency by introducing intermediary libs
> 
> What are your thoughts?

I prefer option a). Because CrossTU is ought to be independent from the analyzer, that is more generic.

martong reviewed (approved):
Looks good!

Szelethus commented:
Shouldn't this be an analyzer config option instead? I am aware of several CTU options that are just that, why is this a frontend option?

martong commented:
> Shouldn't this be an analyzer config option instead? I am aware of several CTU options that are just that, why is this a frontend option?

Yes this should be an analyzer option, but the first attempt somehow created a lib  dependency failure.
So, Endre is working on the second attempt where he will extend the parameters of `CrossTranslationUnitContext::getCrossTUDefinition` like we do with `DisplayCTUProgress`.

martong reviewed (changes requested):
Please extend the parameters of CrossTranslationUnitContext::getCrossTUDefinition like we do with DisplayCTUProgress.

Szelethus commented:
> > Shouldn't this be an analyzer config option instead? I am aware of several CTU options that are just that, why is this a frontend option?
> 
> Yes this should be an analyzer option, but the first attempt somehow created a lib dependency failure.

The reason probably was that my analyzer config changes are not live on clang7 -- the solution would be to leave the option's getter method in-class, I believe.

dkrupp commented:
deadline for this is 31st of March!

gamesh411 committed:
Add analyzer option to limit the number of imported TUs

During CTU analysis of complex projects, the loaded AST-contents of
imported TUs can grow bigger than available system memory. This option
introduces a threshold on the number of TUs to be imported for a single
TU in order to prevent such cases.

gamesh411 commented:
Reorganized parameter handling, added unit test and some documentation.

Szelethus commented:
Sorry for missing out on the discussion. I think adding this as an analyzer-config flag WOULD be preferred on clang 8.0.0, but since we can't list and verify these options in 7.0.0., let's not shoot ourselves in the foot with them.

LGTM regarding option placement, I can't say much about the rest of the patch.

martong commented on include/clang/StaticAnalyzer/Core/AnalyzerOptions.h:
```diff
@@ -186,6 +186,10 @@ class AnalyzerOptions : public RefCountedBase<AnalyzerOptions> {
 
+  /// The maximal amount of translation units that is considered for import
+  /// when inlining functions during CTU analysis.
+  unsigned CTUImportThreshold{100u};
```
I am not sure that aggregate initialization will be compiled without problems on GCC 4.8. Clang master still supports that compiler with C++11, right?
Perhaps ` = 100u;` would be more appropriate.

martong reviewed (approved):
LG! Nice work!
